internal/pkg/istio/config_factory_test.go

```diff
@@ -0,0 +1,306 @@
+// Copyright Red Hat, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the License);
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an AS IS BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package istio
+
+import (
+	"context"
+	"encoding/json"
+	"os"
+	"path/filepath"
+	"reflect"
+	"testing"
+
+	istionetv1alpha3 "istio.io/api/networking/v1alpha3"
+	"istio.io/client-go/pkg/apis/networking/v1alpha3"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes/fake"
+
+	"github.com/openshift-service-mesh/federation/internal/pkg/config"
+	"github.com/openshift-service-mesh/federation/internal/pkg/fds"
+	"github.com/openshift-service-mesh/federation/internal/pkg/informer"
+)
+
+var (
+	exportConfig = config.Federation{
+		MeshPeers: config.MeshPeers{
+			Local: config.Local{
+				Name: "east",
+				ControlPlane: config.ControlPlane{
+					Namespace: "istio-system",
+				},
+				Gateways: config.Gateways{
+					Ingress: config.LocalGateway{
+						Selector: map[string]string{
+							"app": "federation-ingress-gateway",
+						},
+						Port: &config.GatewayPort{
+							Name:   "tls",
+							Number: 443,
+						},
+					},
+				},
+			},
+		},
+		ExportedServiceSet: config.ExportedServiceSet{
+			Rules: []config.Rules{{
+				Type: "LabelSelector",
+				LabelSelectors: []config.LabelSelectors{{
+					MatchLabels: map[string]string{
+						"export": "true",
+					},
+				}},
+			}},
+		},
+	}
+
+	unexportService = &corev1.Service{
+		ObjectMeta: v1.ObjectMeta{Name: "a", Namespace: "ns1"},
+	}
+	exportedService1 = &corev1.Service{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      "b",
+			Namespace: "ns1",
+			Labels: map[string]string{
+				"app":    "b",
+				"export": "true",
+			},
+		},
+		Spec: corev1.ServiceSpec{
+			Ports: []corev1.ServicePort{{
+				Port: 8080,
+			}},
+		},
+	}
+	exportedService2 = &corev1.Service{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      "a",
+			Namespace: "ns2",
+			Labels: map[string]string{
+				"app":    "a",
+				"export": "true",
+			},
+		},
+		Spec: corev1.ServiceSpec{
+			Ports: []corev1.ServicePort{{
+				Port: 9080,
+			}},
+		},
+	}
+)
+
+func TestIngressGateway(t *testing.T) {
+	testCases := []struct {
+		name             string
+		existingServices []*corev1.Service
+		expectedGateway  *v1alpha3.Gateway
+	}{{
+		name:             "found 2 services matching configured label selector",
```

Comment:
nit: the name is a bit confusing for me. it's more about defining 2 exported services as hosts for gateway than finding based on the selector?

Reply:
Copy-paste... I changed this name, I hope it will be clear now.